include/seqan3/range/view/take_until.hpp

```diff
@@ -0,0 +1,440 @@
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+ * \brief Provides seqan3::view::take_until and seqan3::view::take_until_or_throw.
+ */
+
+#pragma once
+
+#include <range/v3/view/take_while.hpp>
+#include <range/v3/algorithm/copy.hpp>
+
+#include <seqan3/core/metafunction/iterator.hpp>
+#include <seqan3/core/metafunction/range.hpp>
+#include <seqan3/io/exception.hpp>
+#include <seqan3/range/view/detail.hpp>
+#include <seqan3/range/detail/inherited_iterator_base.hpp>
+#include <seqan3/std/concept/iterator.hpp>
+#include <seqan3/std/concept/callable.hpp>
+#include <seqan3/std/concept/container.hpp>
+#include <seqan3/std/view/view_all.hpp>
+
+namespace seqan3::detail
+{
+
+// ============================================================================
+//  view_take_until
+// ============================================================================
+
+/*!\brief The type returned by seqan3::view::take_until and seqan3::view::take_until_or_throw.
+ * \tparam urng_t    The type of the underlying ranges, must satisfy seqan3::view::concept.
+ * \tparam or_throw  Whether to throw an exception when the input is exhausted before the end of line is reached.
+ * \implements seqan3::view_concept
+ * \implements seqan3::random_access_range_concept
+ * \ingroup view
+ *
+ * \details
+ *
+ * Note that most members of this class are generated by ranges::view_interface which is not yet documented here.
+ */
+template <view_concept urng_t, typename fun_t, bool or_throw>
+class view_take_until : public ranges::view_interface<view_take_until<urng_t, fun_t, or_throw>>
+{
+private:
+
+    static_assert(invocable_concept<fun_t, reference_t<urng_t>>,
+                  "The functor type for view::take_until must model invocable_concept<fun_t, reference_t<urng_t>>.");
+    static_assert(boolean_concept<std::result_of_t<fun_t&&(reference_t<urng_t>)>>,
+                  "The functor type for view::take_until must return boolean_concept.");
+
+    //!\brief The underlying range.
+    urng_t urange;
```

Comment:
Yes, but we need the private section at the top anyway, because of the type definitions. Adding a second private section just for the sake of it, wouldn't increase readability from my POV.